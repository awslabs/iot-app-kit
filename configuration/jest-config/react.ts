export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
