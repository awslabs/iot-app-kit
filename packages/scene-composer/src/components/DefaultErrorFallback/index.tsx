import React, { FC } from 'react';
import { Header, TextContent } from '@awsui/components-react';

import useLifecycleLogging from '../../logger/react-logger/hooks/useLifecycleLogging';
import { ERROR_MESSAGE_DICT, SceneComposerRuntimeError } from '../../common/errors';
import { StaticLayout } from '../../layouts/StaticLayout';
import CenteredContainer from '../CenteredContainer';

interface ErrorFallbackProps {
  error?: Error;
}

const DefaultErrorFallback: FC<ErrorFallbackProps> = ({ error }) => {
  const log = useLifecycleLogging('DefaultErrorFallback');

  let errorTitle: string;
  let errorMessage: string;

  if (error instanceof SceneComposerRuntimeError) {
    errorTitle = 'Error';
    errorMessage = error.details.message;
    log?.error(errorMessage, error.innerError as Error);
  } else {
    errorTitle = 'Error';
    if (error instanceof Error) {
      errorMessage = error.message;
    } else if (typeof error === 'string') {
      errorMessage = error;
    } else {
      errorMessage = ERROR_MESSAGE_DICT.UNKNOWN_ERROR;
    }
  }

  return (
    <StaticLayout
      showModal={true}
      modalContent={
        <CenteredContainer header={<Header variant='h2'>{errorTitle}</Header>}>
          <TextContent>
            <p>{errorMessage}</p>
          </TextContent>
        </CenteredContainer>
      }
    />
  );
};

export default DefaultErrorFallback;
