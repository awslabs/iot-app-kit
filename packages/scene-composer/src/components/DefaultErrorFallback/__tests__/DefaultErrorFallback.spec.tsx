import * as React from 'react';
import { create } from 'react-test-renderer';

import { ErrorCode, ErrorLevel, SceneComposerRuntimeError } from '../../..';
import DefaultErrorFallback from '..';

jest.mock('../../../layouts/StaticLayout', () => ({
  StaticLayout: 'StaticLayout',
}));

describe('DefaultErrorFallback', () => {
  const sceneComposerRuntimeError = new SceneComposerRuntimeError({
    level: ErrorLevel.ERROR,
    code: ErrorCode.SC_ERROR_LOAD_SCENE,
    message: 'Testing a SceneComposerRuntimeError',
    context: { test: 'test' },
    innerError: new Error('Testing an innerError'),
  });
  const error = new Error('Testing an error');
  const errorString = 'Testing a string' as any;
  const other = {} as any;

  [
    ['SceneComposerRuntimeError', sceneComposerRuntimeError],
    ['Error', error],
    ['String', errorString],
    ['Unknown', other],
  ].forEach((value) => {
    it(`should render correctly with a ${value[0]} type`, () => {
      const container = create(<DefaultErrorFallback error={value[1] as any} />);

      expect(container).toMatchSnapshot();
    });
  });
});
