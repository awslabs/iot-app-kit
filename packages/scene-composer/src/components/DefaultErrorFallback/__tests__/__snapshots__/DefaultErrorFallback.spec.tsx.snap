// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultErrorFallback should render correctly with a Error type 1`] = `
<StaticLayout
  modalContent={
    <Styled(Component)
      header={
        <Unknown
          variant="h2"
        >
          Error
        </Unknown>
      }
    >
      <Unknown>
        <p>
          Testing an error
        </p>
      </Unknown>
    </Styled(Component)>
  }
  showModal={true}
/>
`;

exports[`DefaultErrorFallback should render correctly with a SceneComposerRuntimeError type 1`] = `
<StaticLayout
  modalContent={
    <Styled(Component)
      header={
        <Unknown
          variant="h2"
        >
          Error
        </Unknown>
      }
    >
      <Unknown>
        <p>
          Testing a SceneComposerRuntimeError
        </p>
      </Unknown>
    </Styled(Component)>
  }
  showModal={true}
/>
`;

exports[`DefaultErrorFallback should render correctly with a String type 1`] = `
<StaticLayout
  modalContent={
    <Styled(Component)
      header={
        <Unknown
          variant="h2"
        >
          Error
        </Unknown>
      }
    >
      <Unknown>
        <p>
          Testing a string
        </p>
      </Unknown>
    </Styled(Component)>
  }
  showModal={true}
/>
`;

exports[`DefaultErrorFallback should render correctly with a Unknown type 1`] = `
<StaticLayout
  modalContent={
    <Styled(Component)
      header={
        <Unknown
          variant="h2"
        >
          Error
        </Unknown>
      }
    >
      <Unknown>
        <p>
          An unexpected error occurred.
        </p>
      </Unknown>
    </Styled(Component)>
  }
  showModal={true}
/>
`;
